lib/SILOptimizer/Mandatory/TFPartition.cpp

```diff
@@ -2743,23 +2785,27 @@ static SILValue createHostSend(SILBuilder &B, SILLocation loc, SILValue value,
 
   auto &ctx = B.getASTContext();
   Type scalarValueTy, tensorValueTy;
-  if (!isTensorHandle(value->getType().getASTType())) {
+  if (isTensorHandle(value->getType().getASTType())) {
+    tensorValueTy = value->getType().getASTType();
+    scalarValueTy = getTensorHandleElementType(tensorValueTy);
+  } else {
     assert(createScalarTensorFn);
     // Here scalar type is something like $Builtin.FPIEEE32 -- convert it to an
     // AccelerableByTensorFlow conforming type like Float first, and then create
     // a scalar tensor to send that value.
     scalarValueTy = value->getType().getASTType();
     if (!scalarValueTy->getAs<StructType>()) {
-      // The value must be defined by a struct_extract like:
-      //   %34 = struct_extract %33 : $Float, #Float._value
-      //
-      // In this case we set `value` to the Float operand like %33 above.
-      auto *SEI =
-          castWithDebugInfo<StructExtractInst>(value->getDefiningInstruction());
-      assert(SEI->getFieldNo() == 0);
-      value = SEI->getOperand();
+      assert(value->getDefiningInstruction());
+      auto *typeDecl =
+          getStdlibNumericTypeDeclFromBuiltinType(scalarValueTy, ctx);
+      if (!typeDecl) {
+        value->dump();
+        llvm_unreachable("Unexpected data type when sending a scalar!");
+      }
+      value = wrapInStruct(value, typeDecl, B, loc);
       scalarValueTy = value->getType().getASTType();
     }
+    assert(scalarValueTy->getAs<StructType>());
```

Comment:
`isa<StructType>(scalarValueTy)` is a better fit :)

Reply:
isa is only better if scalarValueTy happens to be a CanType.  If it is a Type that could be sugared, then getAs is (sadly) correct

Reply:
`getASTType()` returns a `CanType`, so `scalarValueTy` can't be sugared, right?

Reply:
Using `isa` here triggers compiler errors:

```
/usr/local/google/home/hongm/ssd_part/git/swift-base/swift/lib/SILOptimizer/Mandatory/TFPartition.cpp:2837:12: error: call to deleted function 'isa'
    assert(isa<StructType>(scalarValueTy));
           ^~~~~~~~~~~~~~~
/usr/include/assert.h:89:5: note: expanded from macro 'assert'
  ((expr)                                                               \
    ^~~~
/usr/local/google/home/hongm/ssd_part/git/swift-base/swift/include/swift/AST/Type.h:517:1: note: candidate function [with X = swift::StructType] has been explicitly deleted
isa(const Type&) = delete; // Use TypeBase::is instead.
^
/usr/local/google/home/hongm/ssd_part/git/swift-base/llvm/include/llvm/Support/Casting.h:142:56: note: candidate function [with X = swift::StructType, Y = swift::Type]
template <class X, class Y> LLVM_NODISCARD inline bool isa(const Y &Val) {
                                                       ^
/usr/local/google/home/hongm/ssd_part/git/swift-base/swift/include/swift/AST/LayoutConstraint.h:305:32: note: candidate function not viable: no known conversion from 'swift::Type' to 'swift::LayoutConstraint' for 1st argument
template <class X> inline bool isa(LayoutConstraint LC) {
                               ^
/usr/local/google/home/hongm/ssd_part/git/swift-base/swift/include/swift/AST/Type.h:527:32: note: candidate function not viable: no known conversion from 'swift::Type' to 'swift::CanType' for 1st argument
template <class X> inline bool isa(CanType type) {
                               ^
/usr/local/google/home/hongm/ssd_part/git/swift-base/swift/include/swift/AST/Type.h:548:13: note: candidate template ignored: could not match 'CanTypeWrapper<type-parameter-0-1>' against 'swift::Type'
inline bool isa(CanTypeWrapper<P> type) {
            ^
```